### PR TITLE
Aggregate metrics for spans

### PR DESCRIPTION
**Pull request by markushi**
## :scroll: Description
Aggregate metrics based on spans: https://develop.sentry.dev/sdk/metrics/#span-aggregation.

image

image

#skip-changelog
-- as the feature hasn't been released yet.

## :bulb: Motivation and Context

Example event: https://sentry-sdks.sentry.io/performance/sentry-android%3A70c95de097d64ca68ebdbbe6f923d625/?openPanel=open&referrer=metrics#span-a39379e81c714978

In order to improve the link between metrics and spans (see [RFC](https://github.com/getsentry/rfcs/blob/main/text/0123-metrics-correlation.md)) we want to aggregate metric emissions when a span is active.

The span summary is grouped by `export_key` (an unique identifier, ignoring the metric tags) and then by `metric_key` (an unique identifier, including metric tags), [the RFC provides a good example](https://github.com/getsentry/rfcs/blob/main/text/0123-metrics-correlation.md#basics) to illustrate why this is practical.

## :green_heart: How did you test it?
Added unit tests, as well as an integration test

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry-android-core/src/test/java/io/sentry/android/core/ManifestMetadataReaderTest.kt [hunk 1]
```diff
@@ -1395,4 +1395,29 @@ class ManifestMetadataReaderTest {
...
+    fun `applyMetadata reads enableMetrics flag to options`() {
+        // Arrange
+        val bundle = bundleOf(ManifestMetadataReader.ENABLE_METRICS to true)
+        val context = fixture.getContext(metaData = bundle)
+
+        // Act
+        ManifestMetadataReader.applyMetadata(context, fixture.options, fixture.buildInfoProvider)
+
+        // Assert
+        assertTrue(fixture.options.isEnableScopePersistence)
```
Comment (stefanosiano) [inline sentry-android-core/src/test/java/io/sentry/android/core/ManifestMetadataReaderTest.kt:1409]:
```suggestion
        assertTrue(fixture.options.isEnableMetrics)
```
😄

File: sentry/src/main/java/io/sentry/Hub.java [hunk 2]
```diff
@@ -955,9 +956,37 @@ private IScope buildLocalScope(
...
+    }
+    return Collections.unmodifiableMap(tags);
+  }
+
+  @Override
+  public @Nullable LocalMetricsAggregator getLocalMetricsAggregator() {
+    if (!options.isEnableSpanLocalMetricAggregation()) {
+      return null;
+    }
+    final @Nullable ISpan span = stack.peek().getScope().getSpan();
```
Comment (stefanosiano) [inline sentry/src/main/java/io/sentry/Hub.java:986]:
```suggestion
    final @Nullable ISpan span = getSpan();
```

File: sentry-samples/sentry-samples-android/src/main/java/io/sentry/samples/android/MetricsActivity.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                        }) {
+                            Text(text = "Distribution")
+                        }
+                        Button(onClick = {
+                            Sentry.metrics().gauge("example.gauge", Random.nextDouble())
+                        }) {
+                            Text(text = "Gauge")
+                        }
+                        Button(onClick = {
+                            Sentry.metrics().set("example.gauge", Random.nextInt())
```
Comment (stefanosiano) [inline sentry-samples/sentry-samples-android/src/main/java/io/sentry/samples/android/MetricsActivity.kt:43]:
```suggestion
                            Sentry.metrics().set("example.set", Random.nextInt())
```

File: sentry/src/main/java/io/sentry/Span.java [hunk 4]
```diff
@@ -44,6 +45,8 @@ public final class Span implements ISpan {
   private final @NotNull Map<String, Object> data = new ConcurrentHashMap<>();
   private final @NotNull Map<String, MeasurementValue> measurements = new ConcurrentHashMap<>();
 
+  private volatile @Nullable LocalMetricsAggregator metricsAggregator = null;
```
Comment (stefanosiano) [inline sentry/src/main/java/io/sentry/Span.java:48]: We have a `LazyEvaluator` you can use here
```suggestion
  private volatile @Nullable LazyEvaluator<LocalMetricsAggregator> metricsAggregatorEvaluator = new LazyEvaluator<>(() -> new LocalMetricsAggregator());
```
and later use `metricsAggregatorEvaluator.getValue()`

Comment (markushi) [reply sentry/src/main/java/io/sentry/Span.java:48]: oh nice, TIL!

File: sentry/src/main/java/io/sentry/Span.java [hunk 5]
```diff
@@ -44,6 +46,15 @@ public final class Span implements ISpan {
...
   private final @NotNull Map<String, MeasurementValue> measurements = new ConcurrentHashMap<>();
 
+  private final @NotNull LazyEvaluator<LocalMetricsAggregator> metricsAggregator =
+      new LazyEvaluator<>(
+          new LazyEvaluator.Evaluator<LocalMetricsAggregator>() {
+            @Override
+            public @NotNull LocalMetricsAggregator evaluate() {
+              return new LocalMetricsAggregator();
+            }
+          });
```
Comment (stefanosiano) [inline sentry/src/main/java/io/sentry/Span.java:56]:
```suggestion
  @SuppressWarnings("Convert2MethodRef") // older AGP versions do not support method references
  private final @NotNull LazyEvaluator<LocalMetricsAggregator> metricsAggregator =
      new LazyEvaluator<>(() -> new LocalMetricsAggregator());
```

File: sentry/src/test/java/io/sentry/SpanTest.kt [hunk 6]
```diff
@@ -486,6 +486,14 @@ class SpanTest {
         assertEquals(1, transaction.root.measurements["test"]!!.value)
     }
 
+    @Test
+    fun `span provides local metrics aggregator instance`() {
+        val span = fixture.getSut()
+        assertNotNull(span.localMetricsAggregator)
+
+        assertSame(span.localMetricsAggregator, span.localMetricsAggregator)
```
Comment (stefanosiano) [inline sentry/src/test/java/io/sentry/SpanTest.kt:494]: is it a mistake?

File: sentry/src/test/java/io/sentry/metrics/MetricsApiTest.kt [hunk 7]
```diff
@@ -206,7 +225,91 @@ class MetricsApiTest {
...
             anyOrNull()
         )
     }
+
+    @Test
+    fun `local aggregator is provided to aggregator`() {
+        val aggregator = mock<IMetricsAggregator>()
+        val localMetricsAggregator = mock<LocalMetricsAggregator>()
+
+        val api = MetricsApi(object : IMetricsInterface {
```
Comment (stefanosiano) [inline sentry/src/test/java/io/sentry/metrics/MetricsApiTest.kt:238]: I would extract this into a `Fixture` class, and the `verify`s into some private methods

**Review by stefanosiano — APPROVED** [review]
Just left few minor comments, but good!

File: sentry/src/test/java/io/sentry/SpanTest.kt (hunk 6 above)
Comment (markushi) [reply sentry/src/test/java/io/sentry/SpanTest.kt:494]: Nope this is intended, it just ensures the getter returns a reference to the same `LocalMetricsAggregator`.
Let me add a comment for this 👍

**Pull request merged**